[FLINK-29093][table] Fix InternalCompilerException in LookupJoinITCase + reset resource counter before each test

SmirAlex opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Fixed unstable test LookupJoinITCase with Full cache


## Brief change log

  - Fixed concurrent code generation of projection object


## Verifying this change

This change is already covered by existing tests, such as LookupJoinITCase.


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes / no / don't know)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented?  not documented)

SmirAlex commented:
Successful pipeline - https://dev.azure.com/asmirnov1/Flink/_build/results?buildId=35&view=results

SmirAlex commented:
Successful pipeline after update - https://dev.azure.com/asmirnov1/Flink/_build/results?buildId=36&view=results

HuangXingBo commented:
cc @PatrickRen

PatrickRen reviewed (commented):
Thanks for the patch @SmirAlex ! I left some comments. Also I think it's necessary to revisit the entire closing process of the full cache. I didn't check thoroughly but it looks like the input format is never closed in the entire flow, which may lead to potential resource leak.

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/keyselector/GenericRowDataKeySelector.java:
```diff
@@ -44,12 +44,16 @@ public GenericRowDataKeySelector(
         this.keySerializer = keySerializer;
     }
 
+    public void open() {
```
I suppose, one of possible causes of issue from the jira ticket is concurrent call of `generatedProjection.newInstance(cl)`. So i extracted it in the separate method, all such calls executes sequentially in one thread

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/LookupFullCache.java:
```diff
@@ -72,6 +72,13 @@ public synchronized void open(Configuration parameters) throws Exception {
                                 } else {
                                     reloadFailCause.addSuppressed(th);
                                 }
+                                try {
```
I used #close to be sure that reload won't happen again after exception. It was possible in case when there are no input records in interval between previous and next reloads. But actually it's not necessary to call #close, I changed the logic, so now flag `isStopped` in `CacheLoader` is being set after encountering an exception .

SmirAlex commented:
Hi @PatrickRen, I updated the PR according to your comments. About closing of InputFormats - they are closed in the finally block of `InputSplitCacheLoadTask#run`. Also I added checks for open-closed counter in all tests with `FullCacheTestInputFormat`.

PatrickRen reviewed (approved):
Thanks for the update @SmirAlex ! LGTM. Could you squash your commits and rebase to the latest master? Also we need another PR to backport it to release-1.16.
